### ISSUE TITLE
Improve case insensitive RegExp unicode performance

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#334]
Case insensitive RegExp compilation normalizes a case sensitive codepoint range into a case insensitive codepoint range by simulating all 16-bit codepoints and seeing what comes out of each conversion. Ranges are coalesced based on the output.

This is done because a single range may split into multiple ranges in case conversion, so it's not enough to simply case normalize the range endpoints. This approach is not ideal, and it won't scale to non-BMP codepoints (but since those are not part of the standard, this is not a big limitation).

There's a concrete performance issue with the current approach because every one of those 65536 codepoint normalization lookups will walk the bit-packed Unicode case conversion ranges.

This can be mitigated in many way:

- Making the case conversion ranges indexed which would improve performance.
- Making the normalization algorithm aware of the Unicode ranges so that a single pass through the normalization ranges (in order) would be enough.
- Reworking the normalization algorithm to be aware of Unicode ranges so that the RegExp compiler didn't need to walk through 65536 codepoints.
- Reworking normalization so that it happens at runtime instead of at compilation time - leading to a different performance tradeoff between compilation and execution.

**Comment by svaarala** [comment]
The workaround in #411 will go into Duktape 1.4 release (it's used by NetSurf).

**Comment by svaarala** [comment]
This issue also affects compilation of http://momentjs.com/ - compilation time is roughly 0,4 seconds my laptop with default options. With `DUK_OPT_REGEXP_CANON_WORKAROUND` enabled (which improves case insensitive regexp compilation performance) it is about 0,025 (~15x faster).

**Comment by svaarala** [comment]
#1616 provides a ~50x speedup with 300 byte footprint impact.

**Comment by fatcerberus** [comment]
So, even more of a speedup than the 128kB lookup table?  Nice.

**Comment by svaarala** [comment]
No, the 128kB lookup table is much faster. But it's actually a separate mechanism:
- The mechanism in this pull speeds up the case insensitive regexp case specifically.
- The 128kB lookup table speeds up per codepoint canonicalization which improves regexps but also other uppercase conversions.
